From e37658deee3b750db2af6d9d99505d18b1141ddf Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Tue, 24 Sep 2024 21:47:52 +0800
Subject: [PATCH 155/338] BACKPORT: OPENEULER: Revert "arm64: head.S:
 Initialise MPAM EL2 registers and disable traps"

hulk inclusion
category: bugfix
bugzilla: https://gitee.com/src-openeuler/kernel/issues/I9OXPO

--------------------------------

This reverts commit a01fda352718f178db2effe5e9d2c1b6a5602356.

If BIOS firmware doesn't enable MPAM function under EL3 environment, while
the hardware has the MPAM ability, that would cause illegal instruction
fault when access MPAM registers.

It should be noted that this will bring potential risks to Kdump. If the
previous kernel has an unhandled MPAM exception interrupt, it would cause
Kdump crash after re-enable interrupt capability.

Fixes: a01fda352718 ("arm64: head.S: Initialise MPAM EL2 registers and disable traps")
Signed-off-by: Zeng Heng <zengheng4@huawei.com>

Link: https://gitee.com/openeuler/kernel/commit/8888a342b9fd6fea79d0cc7884bcecab63157952
[Kexy: Resolved minor conflict in arch/arm64/include/asm/el2_setup.h]
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 arch/arm64/include/asm/el2_setup.h | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/arch/arm64/include/asm/el2_setup.h b/arch/arm64/include/asm/el2_setup.h
index 4ef52d7245bb..b33cd6c0de54 100644
--- a/arch/arm64/include/asm/el2_setup.h
+++ b/arch/arm64/include/asm/el2_setup.h
@@ -249,19 +249,6 @@
 	msr	spsr_el2, x0
 .endm
 
-.macro __init_el2_mpam
-	/* Memory Partitioning And Monitoring: disable EL2 traps */
-	mrs	x1, id_aa64pfr0_el1
-	ubfx	x0, x1, #ID_AA64PFR0_EL1_MPAM_SHIFT, #4
-	cbz	x0, .Lskip_mpam_\@		// skip if no MPAM
-	msr_s	SYS_MPAM2_EL2, xzr		// use the default partition
-						// and disable lower traps
-	mrs_s	x0, SYS_MPAMIDR_EL1
-	tbz	x0, #MPAMIDR_EL1_HAS_HCR_SHIFT, .Lskip_mpam_\@	// skip if no MPAMHCR reg
-	msr_s	SYS_MPAMHCR_EL2, xzr		// clear TRAP_MPAMIDR_EL1 -> EL2
-.Lskip_mpam_\@:
-.endm
-
 /**
  * Initialize EL2 registers to sane values. This should be called early on all
  * cores that were booted in EL2. Note that everything gets initialised as
@@ -279,7 +266,6 @@
 	__init_el2_stage2
 	__init_el2_gicv3
 	__init_el2_hstr
-	__init_el2_mpam
 	__init_el2_nvhe_idregs
 	__init_el2_cptr
 	__init_el2_fgt
-- 
2.48.1

